File: apps/Android/issueskit/src/main/kotlin/com/druware/issueskit/IssuesJSONCoder.kt
package com.druware.issueskit

import com.druware.issueskit.internal.FoundationJson
import com.druware.issueskit.internal.JsonValue
import java.time.Instant
import java.util.Locale
import java.util.UUID
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Reads and writes the `.issues` JSON document format.
 *
 * Output is deliberately diff-friendly, because `.issues` files live in project repositories:
 * keys are sorted, the JSON is pretty-printed, slashes are not escaped, dates are ISO-8601, and
 * the file ends with a newline. Encoding the same model twice therefore produces identical bytes —
 * and, by construction, the same bytes the Apple app produces (see [FoundationJson]).
 *
 * Decoding is tolerant in exactly the ways the format documents: absent keys fall back to their
 * documented defaults at every nesting level, unknown keys are ignored, and unknown enum raw
 * values fall back to that enum's default. The two deliberate exceptions are
 * [Issue.resolutionKind], which decodes to `null` rather than inventing a close reason, and
 * [RemoteLink.provider], which preserves an unrecognized value verbatim because a provider is sync
 * identity. [Relation.issueID] is the one field with no default and its absence is fatal.
 */
object IssuesJSONCoder {

    private val parser = Json

    /** Encodes a model to UTF-8 JSON with a trailing newline. */
    fun encode(model: IssuesDocumentModel): ByteArray = encodeToString(model).toByteArray(Charsets.UTF_8)

    /** Encodes a model to JSON text with a trailing newline. */
    fun encodeToString(model: IssuesDocumentModel): String =
        FoundationJson.render(model.toJson()) + "\n"

    /**
     * Decodes UTF-8 JSON into a model.
     *
     * @throws IssuesError.MissingSchemaVersion when the document declares no `schemaVersion`.
     * @throws IssuesError.UnsupportedSchemaVersion when it was written by a newer build.
     * @throws IssuesError.DecodingFailed when the JSON is malformed or a value has the wrong type.
     */
    fun decode(data: ByteArray): IssuesDocumentModel = decode(data.toString(Charsets.UTF_8))

    /** Decodes JSON text into a model. See [decode]. */
    fun decode(text: String): IssuesDocumentModel {
        val root = try {
            parser.parseToJsonElement(text)
        } catch (error: Exception) {
            throw IssuesError.DecodingFailed(error.message ?: error.toString())
        }
        return documentFrom(root.expectObject("<root>"))
    }

    // MARK: - Decoding

    private fun documentFrom(json: JsonObject): IssuesDocumentModel {
        val version = json.int("schemaVersion") ?: throw IssuesError.MissingSchemaVersion
        if (version > IssuesDocumentModel.SUPPORTED_SCHEMA_VERSION) {
            throw IssuesError.UnsupportedSchemaVersion(version, IssuesDocumentModel.SUPPORTED_SCHEMA_VERSION)
        }
        // Migration seam: only version 1 exists today, so an older version is read as-is. When
        // version 2 lands, branch here — decode the old shape, transform it, and set
        // `schemaVersion` to the migrated value so the next save writes the new shape.

        return IssuesDocumentModel(
            schemaVersion = version,
            project = json.obj("project")?.let(::projectFrom) ?: ProjectInfo(),
            integrations = json.obj("integrations")?.let(::integrationsFrom) ?: IntegrationSettings(),
            labels = json.arr("labels")?.mapObjects("labels", ::labelFrom) ?: emptyList(),
            milestones = json.arr("milestones")?.mapObjects("milestones", ::milestoneFrom) ?: emptyList(),
            people = json.arr("people")?.mapObjects("people", ::personFrom) ?: emptyList(),
            export = json.obj("export")?.let(::exportFrom) ?: ExportSettings(),
            issues = json.arr("issues")?.mapObjects("issues", ::issueFrom) ?: emptyList(),
        )
    }

    private fun projectFrom(json: JsonObject) = ProjectInfo(
        id = json.uuid("id") ?: UUID.randomUUID(),
        name = json.string("name") ?: "",
        summary = json.string("summary") ?: "",
    )

    private fun exportFrom(json: JsonObject) = ExportSettings(
        preambleMarkdown = json.string("preambleMarkdown") ?: ExportSettings.DEFAULT_PREAMBLE_MARKDOWN,
    )

    private fun integrationsFrom(json: JsonObject) = IntegrationSettings(
        github = json.obj("github")?.let { source ->
            GitHubIntegration(
                owner = source.string("owner") ?: "",
                repository = source.string("repository") ?: "",
                defaultLabels = source.stringList("defaultLabels") ?: emptyList(),
                defaultAssignees = source.stringList("defaultAssignees") ?: emptyList(),
                defaultMilestone = source.string("defaultMilestone"),
            )
        },
        azureDevOps = json.obj("azureDevOps")?.let { source ->
            AzureDevOpsIntegration(
                organization = source.string("organization") ?: "",
                project = source.string("project") ?: "",
                team = source.string("team"),
                areaPath = source.string("areaPath"),
                iterationPath = source.string("iterationPath"),
                defaultWorkItemType = source.string("defaultWorkItemType") ?: "Issue",
            )
        },
    )

    private fun labelFrom(json: JsonObject) = LabelDefinition(
        name = json.string("name") ?: "",
        colorHex = json.string("colorHex"),
        description = json.string("description") ?: "",
    )

    private fun milestoneFrom(json: JsonObject) = Milestone(
        name = json.string("name") ?: "",
        dueOn = json.instant("dueOn"),
        isClosed = json.bool("isClosed") ?: false,
    )

    private fun personFrom(json: JsonObject) = Person(
        handle = json.string("handle") ?: "",
        displayName = json.string("displayName") ?: "",
        email = json.string("email") ?: "",
    )

    private fun issueFrom(json: JsonObject): Issue {
        val now = Instant.now()
        return Issue(
            uuid = json.uuid("uuid") ?: UUID.randomUUID(),
            number = json.int("number") ?: 0,
            title = json.string("title") ?: "",
            type = json.string("type")?.let(IssueType::fromRaw) ?: IssueType.DEFAULT,
            priority = json.string("priority")?.let(IssuePriority::fromRaw) ?: IssuePriority.DEFAULT,
            status = json.string("status")?.let(IssueStatus::fromRaw) ?: IssueStatus.DEFAULT,
            // Never guesses: an unrecognized close reason would misreport why work stopped.
            resolutionKind = json.string("resolutionKind")?.let(ResolutionKind::fromRaw),
            labels = json.stringList("labels") ?: emptyList(),
            assignees = json.stringList("assignees") ?: emptyList(),
            milestone = json.string("milestone"),
            area = json.string("area") ?: "",
            estimate = json.double("estimate"),
            reportedBy = json.string("reportedBy") ?: "",
            reported = json.instant("reported") ?: now,
            createdAt = json.instant("createdAt") ?: now,
            updatedAt = json.instant("updatedAt") ?: now,
            closedAt = json.instant("closedAt"),
            description = json.string("description") ?: "",
            stepsToReproduce = json.stringList("stepsToReproduce") ?: emptyList(),
            environment = json.string("environment") ?: "",
            notes = json.string("notes") ?: "",
            resolution = json.string("resolution") ?: "",
            comments = json.arr("comments")?.mapObjects("comments", ::commentFrom) ?: emptyList(),
            relations = json.arr("relations")?.mapObjects("relations", ::relationFrom) ?: emptyList(),
            remoteLinks = json.arr("remoteLinks")?.mapObjects("remoteLinks", ::remoteLinkFrom) ?: emptyList(),
        )
    }

    private fun commentFrom(json: JsonObject) = Comment(
        id = json.uuid("id") ?: UUID.randomUUID(),
        author = json.string("author") ?: "",
        createdAt = json.instant("createdAt") ?: Instant.now(),
        body = json.string("body") ?: "",
    )

    private fun relationFrom(json: JsonObject) = Relation(
        kind = json.string("kind")?.let(RelationKind::fromRaw) ?: RelationKind.DEFAULT,
        // The one field with no default — a relation pointing nowhere is not a relation.
        issueID = json.uuid("issueID")
            ?: throw IssuesError.DecodingFailed("relations[] entry is missing the required \"issueID\" key"),
    )

    private fun remoteLinkFrom(json: JsonObject) = RemoteLink(
        // Absence takes the default; an unrecognized value present in the file is preserved.
        provider = json.string("provider")?.let(RemoteProvider::fromRaw) ?: RemoteProvider.Github,
        identifier = json.string("identifier") ?: "",
        url = json.string("url"),
        lastSyncedAt = json.instant("lastSyncedAt"),
        remoteUpdatedAt = json.instant("remoteUpdatedAt"),
    )

    // MARK: - Encoding

    private fun IssuesDocumentModel.toJson() = obj {
        put("schemaVersion", schemaVersion)
        put("project", project.toJson())
        put("integrations", integrations.toJson())
        put("labels", labels.map { it.toJson() })
        put("milestones", milestones.map { it.toJson() })
        put("people", people.map { it.toJson() })
        put("export", export.toJson())
        put("issues", issues.map { it.toJson() })
    }

    private fun ProjectInfo.toJson() = obj {
        put("id", id)
        put("name", name)
        put("summary", summary)
    }

    private fun ExportSettings.toJson() = obj { put("preambleMarkdown", preambleMarkdown) }

    private fun IntegrationSettings.toJson() = obj {
        putIfPresent("github", github?.toJson())
        putIfPresent("azureDevOps", azureDevOps?.toJson())
    }

    private fun GitHubIntegration.toJson() = obj {
        put("owner", owner)
        put("repository", repository)
        put("defaultLabels", defaultLabels)
        put("defaultAssignees", defaultAssignees)
        putIfPresent("defaultMilestone", defaultMilestone)
    }

    private fun AzureDevOpsIntegration.toJson() = obj {
        put("organization", organization)
        put("project", project)
        putIfPresent("team", team)
        putIfPresent("areaPath", areaPath)
        putIfPresent("iterationPath", iterationPath)
        put("defaultWorkItemType", defaultWorkItemType)
    }

    private fun LabelDefinition.toJson() = obj {
        put("name", name)
        putIfPresent("colorHex", colorHex)
        put("description", description)
    }

    private fun Milestone.toJson() = obj {
        put("name", name)
        putIfPresent("dueOn", dueOn)
        put("isClosed", isClosed)
    }

    private fun Person.toJson() = obj {
        put("handle", handle)
        put("displayName", displayName)
        put("email", email)
    }

    private fun Issue.toJson() = obj {
        put("uuid", uuid)
        put("number", number)
        put("title", title)
        put("type", type.raw)
        put("priority", priority.raw)
        put("status", status.raw)
        putIfPresent("resolutionKind", resolutionKind?.raw)
        put("labels", labels)
        put("assignees", assignees)
        putIfPresent("milestone", milestone)
        put("area", area)
        putIfPresent("estimate", estimate)
        put("reportedBy", reportedBy)
        put("reported", reported)
        put("createdAt", createdAt)
        put("updatedAt", updatedAt)
        putIfPresent("closedAt", closedAt)
        put("description", description)
        put("stepsToReproduce", stepsToReproduce)
        put("environment", environment)
        put("notes", notes)
        put("resolution", resolution)
        put("comments", comments.map { it.toJson() })
        put("relations", relations.map { it.toJson() })
        put("remoteLinks", remoteLinks.map { it.toJson() })
    }

    private fun Comment.toJson() = obj {
        put("id", id)
        put("author", author)
        put("createdAt", createdAt)
        put("body", body)
    }

    private fun Relation.toJson() = obj {
        put("kind", kind.raw)
        put("issueID", issueID)
    }

    private fun RemoteLink.toJson() = obj {
        put("provider", provider.raw)
        put("identifier", identifier)
        putIfPresent("url", url)
        putIfPresent("lastSyncedAt", lastSyncedAt)
        putIfPresent("remoteUpdatedAt", remoteUpdatedAt)
    }

    // MARK: - Encoding helpers

    private class ObjectScope {
        val entries = LinkedHashMap<String, JsonValue>()

        fun put(key: String, value: String) { entries[key] = JsonValue.Str(value) }
        fun put(key: String, value: Int) { entries[key] = JsonValue.Num(value.toString()) }
        fun put(key: String, value: Boolean) { entries[key] = JsonValue.Bool(value) }
        fun put(key: String, value: UUID) { entries[key] = JsonValue.Str(value.toString().uppercase(Locale.ROOT)) }
        fun put(key: String, value: Instant) { entries[key] = JsonValue.Str(IssueTimestamp.format(value)) }
        fun put(key: String, value: JsonValue) { entries[key] = value }

        @JvmName("putStrings")
        fun put(key: String, values: List<String>) {
            entries[key] = JsonValue.Arr(values.map { JsonValue.Str(it) })
        }

        @JvmName("putObjects")
        fun put(key: String, values: List<JsonValue>) { entries[key] = JsonValue.Arr(values) }

        /** Mirrors Swift's `encodeIfPresent`: a null optional writes no key at all. */
        fun putIfPresent(key: String, value: String?) { if (value != null) put(key, value) }

        @JvmName("putInstantIfPresent")
        fun putIfPresent(key: String, value: Instant?) { if (value != null) put(key, value) }

        @JvmName("putDoubleIfPresent")
        fun putIfPresent(key: String, value: Double?) {
            if (value != null) entries[key] = JsonValue.Num(FoundationJson.numberLiteral(value))
        }

        @JvmName("putValueIfPresent")
        fun putIfPresent(key: String, value: JsonValue?) { if (value != null) put(key, value) }
    }

    private inline fun obj(build: ObjectScope.() -> Unit): JsonValue.Obj =
        JsonValue.Obj(ObjectScope().apply(build).entries)

    // MARK: - Decoding helpers

    private fun JsonElement.expectObject(path: String): JsonObject =
        this as? JsonObject ?: throw IssuesError.DecodingFailed("expected an object at \"$path\"")

    private fun JsonObject.value(key: String): JsonElement? =
        this[key]?.takeUnless { it is JsonNull }

    private fun JsonObject.primitive(key: String): JsonPrimitive? =
        value(key)?.let {
            it as? JsonPrimitive ?: throw IssuesError.DecodingFailed("\"$key\" is not a scalar value")
        }

    private fun JsonObject.string(key: String): String? = primitive(key)?.let {
        if (!it.isString) throw IssuesError.DecodingFailed("\"$key\" is not a string")
        it.content
    }

    private fun JsonObject.int(key: String): Int? = primitive(key)?.let {
        it.content.toIntOrNull() ?: throw IssuesError.DecodingFailed("\"$key\" is not an integer")
    }

    private fun JsonObject.double(key: String): Double? = primitive(key)?.let {
        it.content.toDoubleOrNull() ?: throw IssuesError.DecodingFailed("\"$key\" is not a number")
    }

    private fun JsonObject.bool(key: String): Boolean? = primitive(key)?.let {
        when (it.content) {
            "true" -> true
            "false" -> false
            else -> throw IssuesError.DecodingFailed("\"$key\" is not a boolean")
        }
    }

    private fun JsonObject.uuid(key: String): UUID? = string(key)?.let {
        try {
            UUID.fromString(it)
        } catch (_: IllegalArgumentException) {
            throw IssuesError.DecodingFailed("\"$key\" is not a UUID: $it")
        }
    }

    private fun JsonObject.instant(key: String): Instant? = string(key)?.let {
        IssueTimestamp.parse(it) ?: throw IssuesError.DecodingFailed("\"$key\" is not an ISO-8601 date: $it")
    }

    private fun JsonObject.obj(key: String): JsonObject? = value(key)?.expectObject(key)

    private fun JsonObject.arr(key: String): JsonArray? = value(key)?.let {
        it as? JsonArray ?: throw IssuesError.DecodingFailed("\"$key\" is not an array")
    }

    private fun JsonObject.stringList(key: String): List<String>? = arr(key)?.map { element ->
        val primitive = element as? JsonPrimitive
        if (primitive == null || !primitive.isString) {
            throw IssuesError.DecodingFailed("\"$key\" contains a non-string element")
        }
        primitive.content
    }

    private fun <T> JsonArray.mapObjects(path: String, transform: (JsonObject) -> T): List<T> =
        map { transform(it.expectObject(path)) }
}

File: apps/Windows/src/IssuesKit/Json/IssuesJsonCoder.cs
using System.Globalization;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace IssuesKit.Json;

/// <summary>
/// Reads and writes the <c>.issues</c> JSON document format.
/// </summary>
/// <remarks>
/// Output is deliberately diff-friendly, because <c>.issues</c> files live in project
/// repositories: keys are sorted, the JSON is pretty-printed, slashes are not escaped, dates are
/// ISO-8601, and the file ends with a newline. Encoding the same model twice therefore produces
/// identical bytes — and the same bytes the macOS and iOS builds produce, so a document shared
/// between platforms does not churn.
/// <para>
/// Decoding is forward- and backward-tolerant: absent keys fall back to the documented default,
/// unknown keys are ignored, and unknown enum values fall back to that enum's default — with two
/// deliberate exceptions, <c>resolutionKind</c> and <c>remoteLinks[].provider</c>.
/// </para>
/// </remarks>
public static class IssuesJsonCoder
{
    private static readonly UTF8Encoding Utf8WithoutBom = new(encoderShouldEmitUTF8Identifier: false);

    // MARK: - Encoding

    /// <summary>Encodes a model to UTF-8 JSON with a trailing newline.</summary>
    /// <exception cref="EncodingFailedException">The model could not be rendered.</exception>
    public static byte[] Encode(IssuesDocumentModel model)
    {
        ArgumentNullException.ThrowIfNull(model);

        try
        {
            return Utf8WithoutBom.GetBytes(EncodeToString(model));
        }
        catch (Exception error) when (error is not IssuesException)
        {
            throw new EncodingFailedException(error.Message, error);
        }
    }

    /// <summary>Encodes a model to a JSON string with a trailing newline.</summary>
    /// <exception cref="EncodingFailedException">The model could not be rendered.</exception>
    public static string EncodeToString(IssuesDocumentModel model)
    {
        ArgumentNullException.ThrowIfNull(model);

        try
        {
            return SwiftJson.Render(EncodeDocument(model)) + "\n";
        }
        catch (Exception error) when (error is not IssuesException)
        {
            throw new EncodingFailedException(error.Message, error);
        }
    }

    private static JsonValue EncodeDocument(IssuesDocumentModel model) =>
        new JsonObjectBuilder()
            .Set("export", EncodeExport(model.Export))
            .Set("integrations", EncodeIntegrations(model.Integrations))
            .Set("issues", model.Issues, EncodeIssue)
            .Set("labels", model.Labels, EncodeLabel)
            .Set("milestones", model.Milestones, EncodeMilestone)
            .Set("people", model.People, EncodePerson)
            .Set("project", EncodeProject(model.Project))
            .Set("schemaVersion", model.SchemaVersion)
            .Build();

    private static JsonValue EncodeExport(ExportSettings export) =>
        new JsonObjectBuilder()
            .Set("preambleMarkdown", export.PreambleMarkdown)
            .Build();

    private static JsonValue EncodeProject(ProjectInfo project) =>
        new JsonObjectBuilder()
            .Set("id", project.Id)
            .Set("name", project.Name)
            .Set("summary", project.Summary)
            .Build();

    private static JsonValue EncodeIntegrations(IntegrationSettings integrations)
    {
        var builder = new JsonObjectBuilder();
        if (integrations.AzureDevOps is { } azure)
        {
            builder.Set("azureDevOps", new JsonObjectBuilder()
                .SetIfPresent("areaPath", azure.AreaPath)
                .Set("defaultWorkItemType", azure.DefaultWorkItemType)
                .SetIfPresent("iterationPath", azure.IterationPath)
                .Set("organization", azure.Organization)
                .Set("project", azure.Project)
                .SetIfPresent("team", azure.Team)
                .Build());
        }

        if (integrations.GitHub is { } github)
        {
            builder.Set("github", new JsonObjectBuilder()
                .Set("defaultAssignees", github.DefaultAssignees)
                .Set("defaultLabels", github.DefaultLabels)
                .SetIfPresent("defaultMilestone", github.DefaultMilestone)
                .Set("owner", github.Owner)
                .Set("repository", github.Repository)
                .Build());
        }

        return builder.Build();
    }

    private static JsonValue EncodeLabel(LabelDefinition label) =>
        new JsonObjectBuilder()
            .SetIfPresent("colorHex", label.ColorHex)
            .Set("description", label.Description)
            .Set("name", label.Name)
            .Build();

    private static JsonValue EncodeMilestone(Milestone milestone) =>
        new JsonObjectBuilder()
            .SetIfPresent("dueOn", milestone.DueOn)
            .Set("isClosed", milestone.IsClosed)
            .Set("name", milestone.Name)
            .Build();

    private static JsonValue EncodePerson(Person person) =>
        new JsonObjectBuilder()
            .Set("displayName", person.DisplayName)
            .Set("email", person.Email)
            .Set("handle", person.Handle)
            .Build();

    private static JsonValue EncodeIssue(Issue issue) =>
        new JsonObjectBuilder()
            .Set("area", issue.Area)
            .Set("assignees", issue.Assignees)
            .SetIfPresent("closedAt", issue.ClosedAt)
            .Set("comments", issue.Comments, EncodeComment)
            .Set("createdAt", issue.CreatedAt)
            .Set("description", issue.Description)
            .Set("environment", issue.Environment)
            .SetIfPresent("estimate", issue.Estimate)
            .Set("labels", issue.Labels)
            .SetIfPresent("milestone", issue.Milestone)
            .Set("notes", issue.Notes)
            .Set("number", issue.Number)
            .Set("priority", issue.Priority.RawValue())
            .Set("relations", issue.Relations, EncodeRelation)
            .Set("remoteLinks", issue.RemoteLinks, EncodeRemoteLink)
            .Set("reported", issue.Reported)
            .Set("reportedBy", issue.ReportedBy)
            .Set("resolution", issue.Resolution)
            .SetIfPresent("resolutionKind", issue.ResolutionKind?.RawValue())
            .Set("status", issue.Status.RawValue())
            .Set("stepsToReproduce", issue.StepsToReproduce)
            .Set("title", issue.Title)
            .Set("type", issue.Type.RawValue())
            .Set("updatedAt", issue.UpdatedAt)
            .Set("uuid", issue.Uuid)
            .Build();

    private static JsonValue EncodeComment(Comment comment) =>
        new JsonObjectBuilder()
            .Set("author", comment.Author)
            .Set("body", comment.Body)
            .Set("createdAt", comment.CreatedAt)
            .Set("id", comment.Id)
            .Build();

    private static JsonValue EncodeRelation(Relation relation) =>
        new JsonObjectBuilder()
            .Set("issueID", relation.IssueId)
            .Set("kind", relation.Kind.RawValue())
            .Build();

    private static JsonValue EncodeRemoteLink(RemoteLink link) =>
        new JsonObjectBuilder()
            .Set("identifier", link.Identifier)
            .SetIfPresent("lastSyncedAt", link.LastSyncedAt)
            .Set("provider", link.Provider.RawValue)
            .SetIfPresent("remoteUpdatedAt", link.RemoteUpdatedAt)
            .SetIfPresent("url", link.Url)
            .Build();

    // MARK: - Decoding

    /// <summary>Decodes UTF-8 JSON into a model.</summary>
    /// <exception cref="MissingSchemaVersionException">The document declares no schema version.</exception>
    /// <exception cref="UnsupportedSchemaVersionException">The document was written by a newer build.</exception>
    /// <exception cref="DecodingFailedException">The JSON is malformed or a value has the wrong type.</exception>
    public static IssuesDocumentModel Decode(ReadOnlySpan<byte> data) =>
        Decode(Utf8WithoutBom.GetString(data));

    /// <inheritdoc cref="Decode(ReadOnlySpan{byte})"/>
    public static IssuesDocumentModel Decode(string json)
    {
        JsonNode? root;
        try
        {
            root = JsonNode.Parse(json);
        }
        catch (JsonException error)
        {
            throw new DecodingFailedException(error.Message, error);
        }

        if (root is not JsonObject document)
        {
            throw new DecodingFailedException("The top level of the document is not a JSON object.");
        }

        if (document["schemaVersion"] is not { } versionNode)
        {
            throw new MissingSchemaVersionException();
        }

        var version = AsInt(versionNode, "schemaVersion");
        if (version > IssuesDocumentModel.SupportedSchemaVersion)
        {
            throw new UnsupportedSchemaVersionException(version, IssuesDocumentModel.SupportedSchemaVersion);
        }

        // Migration seam: only version 1 exists today, so an older version is read as-is. When
        // version 2 lands, branch here — decode the old shape, transform it, and set SchemaVersion
        // to the migrated value so the next save writes the new shape.

        return new IssuesDocumentModel
        {
            SchemaVersion = version,
            Project = DecodeProject(Object(document, "project")),
            Integrations = DecodeIntegrations(Object(document, "integrations")),
            Labels = DecodeList(document, "labels", DecodeLabel),
            Milestones = DecodeList(document, "milestones", DecodeMilestone),
            People = DecodeList(document, "people", DecodePerson),
            Export = DecodeExport(Object(document, "export")),
            Issues = DecodeList(document, "issues", DecodeIssue)
        };
    }

    private static ProjectInfo DecodeProject(JsonObject? node) => node is null
        ? new ProjectInfo()
        : new ProjectInfo
        {
            Id = GuidOr(node, "id", System.Guid.NewGuid()),
            Name = StringOr(node, "name", string.Empty),
            Summary = StringOr(node, "summary", string.Empty)
        };

    private static ExportSettings DecodeExport(JsonObject? node) => node is null
        ? new ExportSettings()
        : new ExportSettings
        {
            PreambleMarkdown = StringOr(node, "preambleMarkdown", ExportSettings.DefaultPreambleMarkdown)
        };

    private static IntegrationSettings DecodeIntegrations(JsonObject? node)
    {
        if (node is null)
        {
            return new IntegrationSettings();
        }

        var settings = new IntegrationSettings();

        if (Object(node, "github") is { } github)
        {
            settings.GitHub = new GitHubIntegration
            {
                Owner = StringOr(github, "owner", string.Empty),
                Repository = StringOr(github, "repository", string.Empty),
                DefaultLabels = StringList(github, "defaultLabels"),
                DefaultAssignees = StringList(github, "defaultAssignees"),
                DefaultMilestone = StringOrNull(github, "defaultMilestone")
            };
        }

        if (Object(node, "azureDevOps") is { } azure)
        {
            settings.AzureDevOps = new AzureDevOpsIntegration
            {
                Organization = StringOr(azure, "organization", string.Empty),
                Project = StringOr(azure, "project", string.Empty),
                Team = StringOrNull(azure, "team"),
                AreaPath = StringOrNull(azure, "areaPath"),
                IterationPath = StringOrNull(azure, "iterationPath"),
                DefaultWorkItemType = StringOr(azure, "defaultWorkItemType", "Issue")
            };
        }

        return settings;
    }

    private static LabelDefinition DecodeLabel(JsonObject node) => new()
    {
        Name = StringOr(node, "name", string.Empty),
        ColorHex = StringOrNull(node, "colorHex"),
        Description = StringOr(node, "description", string.Empty)
    };

    private static Milestone DecodeMilestone(JsonObject node) => new()
    {
        Name = StringOr(node, "name", string.Empty),
        DueOn = DateOrNull(node, "dueOn"),
        IsClosed = BoolOr(node, "isClosed", false)
    };

    private static Person DecodePerson(JsonObject node) => new()
    {
        Handle = StringOr(node, "handle", string.Empty),
        DisplayName = StringOr(node, "displayName", string.Empty),
        Email = StringOr(node, "email", string.Empty)
    };

    private static Issue DecodeIssue(JsonObject node)
    {
        var now = DateTimeOffset.UtcNow;
        return new Issue
        {
            Uuid = GuidOr(node, "uuid", System.Guid.NewGuid()),
            Number = IntOr(node, "number", 0),
            Title = StringOr(node, "title", string.Empty),
            Type = IssueEnums.IssueTypeFromRaw(StringOrNull(node, "type")),
            Priority = IssueEnums.IssuePriorityFromRaw(StringOrNull(node, "priority")),
            Status = IssueEnums.IssueStatusFromRaw(StringOrNull(node, "status")),
            // An unrecognized resolution decodes to null rather than a default.
            ResolutionKind = IssueEnums.ResolutionKindFromRaw(StringOrNull(node, "resolutionKind")),
            Labels = StringList(node, "labels"),
            Assignees = StringList(node, "assignees"),
            Milestone = StringOrNull(node, "milestone"),
            Area = StringOr(node, "area", string.Empty),
            Estimate = DoubleOrNull(node, "estimate"),
            ReportedBy = StringOr(node, "reportedBy", string.Empty),
            Reported = DateOrNull(node, "reported") ?? now,
            CreatedAt = DateOrNull(node, "createdAt") ?? now,
            UpdatedAt = DateOrNull(node, "updatedAt") ?? now,
            ClosedAt = DateOrNull(node, "closedAt"),
            Description = StringOr(node, "description", string.Empty),
            StepsToReproduce = StringList(node, "stepsToReproduce"),
            Environment = StringOr(node, "environment", string.Empty),
            Notes = StringOr(node, "notes", string.Empty),
            Resolution = StringOr(node, "resolution", string.Empty),
            Comments = DecodeList(node, "comments", DecodeComment),
            Relations = DecodeList(node, "relations", DecodeRelation),
            RemoteLinks = DecodeList(node, "remoteLinks", DecodeRemoteLink)
        };
    }

    private static Comment DecodeComment(JsonObject node) => new()
    {
        Id = GuidOr(node, "id", System.Guid.NewGuid()),
        Author = StringOr(node, "author", string.Empty),
        CreatedAt = DateOrNull(node, "createdAt") ?? DateTimeOffset.UtcNow,
        Body = StringOr(node, "body", string.Empty)
    };

    /// <summary>
    /// <c>issueID</c> is required — a relation that points nowhere is not a relation.
    /// </summary>
    private static Relation DecodeRelation(JsonObject node)
    {
        if (node["issueID"] is not { } target)
        {
            throw new DecodingFailedException("A relation is missing its required \"issueID\".");
        }

        return new Relation(
            IssueEnums.RelationKindFromRaw(StringOrNull(node, "kind")),
            ParseGuid(AsString(target, "issueID"), "issueID"));
    }

    /// <summary>
    /// An unrecognized provider is preserved verbatim; it is never coerced to a known provider.
    /// </summary>
    private static RemoteLink DecodeRemoteLink(JsonObject node) => new()
    {
        Provider = new RemoteProvider(StringOrNull(node, "provider") ?? RemoteProvider.GitHub.RawValue),
        Identifier = StringOr(node, "identifier", string.Empty),
        Url = UriOrNull(node, "url"),
        LastSyncedAt = DateOrNull(node, "lastSyncedAt"),
        RemoteUpdatedAt = DateOrNull(node, "remoteUpdatedAt")
    };

    // MARK: - Tolerant readers

    private static JsonObject? Object(JsonObject parent, string key) => parent[key] as JsonObject;

    private static List<T> DecodeList<T>(JsonObject parent, string key, Func<JsonObject, T> decode)
    {
        if (parent[key] is not JsonArray array)
        {
            return [];
        }

        var results = new List<T>(array.Count);
        foreach (var item in array)
        {
            if (item is JsonObject element)
            {
                results.Add(decode(element));
            }
        }

        return results;
    }

    private static List<string> StringList(JsonObject parent, string key)
    {
        if (parent[key] is not JsonArray array)
        {
            return [];
        }

        var results = new List<string>(array.Count);
        foreach (var item in array)
        {
            if (item is not null)
            {
                results.Add(AsString(item, key));
            }
        }

        return results;
    }

    private static string StringOr(JsonObject parent, string key, string fallback) =>
        parent[key] is { } node ? AsString(node, key) : fallback;

    private static string? StringOrNull(JsonObject parent, string key) =>
        parent[key] is { } node ? AsString(node, key) : null;

    private static int IntOr(JsonObject parent, string key, int fallback) =>
        parent[key] is { } node ? AsInt(node, key) : fallback;

    private static bool BoolOr(JsonObject parent, string key, bool fallback)
    {
        if (parent[key] is not { } node)
        {
            return fallback;
        }

        try
        {
            return node.GetValue<bool>();
        }
        catch (Exception error) when (error is InvalidOperationException or FormatException)
        {
            throw new DecodingFailedException($"\"{key}\" is not a boolean.", error);
        }
    }

    private static double? DoubleOrNull(JsonObject parent, string key)
    {
        if (parent[key] is not { } node)
        {
            return null;
        }

        try
        {
            return node.GetValue<double>();
        }
        catch (Exception error) when (error is InvalidOperationException or FormatException)
        {
            throw new DecodingFailedException($"\"{key}\" is not a number.", error);
        }
    }

    private static Guid GuidOr(JsonObject parent, string key, Guid fallback) =>
        parent[key] is { } node ? ParseGuid(AsString(node, key), key) : fallback;

    private static DateTimeOffset? DateOrNull(JsonObject parent, string key)
    {
        if (parent[key] is not { } node)
        {
            return null;
        }

        var text = AsString(node, key);
        if (!DateTimeOffset.TryParse(
                text,
                CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                out var value))
        {
            throw new DecodingFailedException($"\"{key}\" is not an ISO-8601 date: \"{text}\".");
        }

        return value;
    }

    /// <summary>A malformed URL is dropped rather than failing the decode, matching the reference build.</summary>
    private static Uri? UriOrNull(JsonObject parent, string key)
    {
        if (parent[key] is not { } node)
        {
            return null;
        }

        return Uri.TryCreate(AsString(node, key), UriKind.RelativeOrAbsolute, out var uri) ? uri : null;
    }

    private static string AsString(JsonNode node, string key)
    {
        try
        {
            return node.GetValue<string>();
        }
        catch (Exception error) when (error is InvalidOperationException or FormatException)
        {
            throw new DecodingFailedException($"\"{key}\" is not a string.", error);
        }
    }

    private static int AsInt(JsonNode node, string key)
    {
        try
        {
            return node.GetValue<int>();
        }
        catch (Exception error) when (error is InvalidOperationException or FormatException)
        {
            throw new DecodingFailedException($"\"{key}\" is not an integer.", error);
        }
    }

    private static Guid ParseGuid(string text, string key)
    {
        if (!System.Guid.TryParse(text, out var value))
        {
            throw new DecodingFailedException($"\"{key}\" is not a UUID: \"{text}\".");
        }

        return value;
    }
}
